morebits.js

```diff
@@ -1152,6 +1152,52 @@ Morebits.array = {
 	}
 };
 
+/**
+ * select2 doesn't natively support highlightening of search hits so we implement that here
+ * @param {jQuery} $select
+ * @param {boolean} [hasOptionGroups=false] - does the select menu have optgroups?
+ * @requires select2
+ */
+Morebits.select2SearchHighlights = function select2SearchHighlights($select, hasOptionGroups) {
+	$select.on('select2:open', function() {
+		$('.select2-search__field')[0].addEventListener('keyup', function() {
+			var $ul = $('.select2-results__options');
+			$ul.find('.search-hit').each(function(_, e) {
+				var li_element = e.parentElement;
+				// This would convert <li>Hello <span class=search-hit>wo</span>rld</li>
+				// to <label>Hello world</label>
+				li_element.innerHTML = li_element.textContent;
+			});
+
+			if (this.value) {
+				var searchString = this.value;
+				var searchRegex = new RegExp(mw.RegExp.escape(searchString), 'i');
+
+				var selectorString;
+				// the odd way in which select2 organises the ul necessitates this hack
+				if (hasOptionGroups) {
+					selectorString = 'li > strong, li > ul > li';
```

Comment:
> I guess the thinking is that since it will show up as a match, users should know why?

Yes. wouldn't it be odd if a optgroup title shows up in the result without a highlight, and there are no options within that optgroup?

Reply:
Maybe?  I find it distracting to have those highlighted as if I can select them.  They're not really proper search results after all.

Reply:
Eh, I think you're probably right.  If you searching for "disruptive" in "common warnings" highlights the result, not common, which is good and normal, but typing "common" gives no results, so, yeah, I suppose it'd be weird.

Reply:
How crazy is something like

```javascript
$('li[role="group"]').each(function (idx, group) {
    if (!$(group).find('li[role="option"]').length) {
        $(group).hide();
    }
});
```

To hide empty optgroups?

Reply:
Actually, it's not too bad with a matcher, what do you think of the below?  Inspired by https://github.com/select2/select2/issues/3034#issuecomment-338286741, it removes optgroups with no options from the result.

```javascript
matcher: function(params, data) {
	var original_matcher = $.fn.select2.defaults.defaults.matcher;
	var result = original_matcher(params, data);
	if (result && (!result.children || !result.children.length)) {
		result = null;
	}
	return result;
}
```

Reply:
Actually the code doesn't fully replicate chosen. I just realised that what chosen does is that if the search term matches in the optgroup name, _all_ options in that optgroup are shown. That seems ideal. I think _this_ is the behaviour we should be emulating. Should be doable with a custom matcher.

Reply:
Right you are!  I actually don't like that behavior, but I think it's probably expected.  Pretty sure that was the original intent of the comment I linked above, should work just fine?

```javascript
matcher: function(params, data) {
	var original_matcher = $.fn.select2.defaults.defaults.matcher;
	var result = original_matcher(params, data);
	if (result && data.children && result.children && data.children.length != result.children.length) {
		result.children = data.children;
	}
	return result;
}
```

Reply:
not quite, that ends up showing everything in the optgroup if any option in the optgroup matches.